manifestoes/extensions.rst

```diff
@@ -0,0 +1,1777 @@
+with ``-fdefer-type-errors``, GHC is perfectly capable of making sense of a
+program with type errors. With ``-XNoDeriveFunctor``, GHC is similarly capable
+of making sense of a program with ``deriving Functor``. Yet in both of these
+scenarios (and many more), GHC issues an error.
+
+I thus argue that the line between warnings and (some) errors is blurry. We
+can imagine a spectrum of code problems going from spelling of identifiers
+through the kinds of problems identified by Hlint through GHC's warnings through
+features enabled by GHC extensions through features enabled by flags like ``-fdefer-type-errors``
+through errors GHC simply cannot recover from (like parser errors). Embodied
+in the current way these potential problems are distributed between different
+reporting mechanisms are value judgments on what the GHC (and other tool) developers
+think about the problems at hand. It is thus reasonable to expect people to differ
+on these points, as with any value judgments. My point here is simply calling
+something an "error" or a "warning" or some other problem is not very informative.
+Accordingly, we should strive for a *uniform treatment* of these potential problems;
+doing so would simplify Haskell's user interface.
+
+Let's also reflect on why we have warnings at all. I claim the usefulness of
+warnings depends on what our user is trying to accomplish. Here are some
+possibilities:
+
+1. The user is actively developing the code in question.
+2. The user is compiling the code in question in order to use the compilation
+   output (as a library or an application); the user is not a developer of the
+   code.
+3. The user is running CI and/or publishing their work.
+
+In scenario (1), warnings are great. They give the user information they might
+use to improve their code. Indeed warnings are better than errors for them, because
+perhaps their focus at the moment is on a different part of the code than where
+the warning arose, or perhaps the user is experimenting and is happy for e.g. an
+import to be unused.
+
+In scenario (2), warnings are mostly useless. The user doesn't care about the
+code being compiled, and doesn't need to know that it has a few unused variables.
+The only reason that warnings aren't completely useless here is that conscientious
+users might reach out to library authors to tell them that their code is warning.
+
+In scenario (3), the fact that warnings do not stop compilation is actively harmful. When going through final checks
+before merging a pull request or posting on Hackage, an unresolved warning is a
+potential problem in code that the programmer might have missed. This might be
+disastrous.
+
+Happily, we can accommodate all of these usage scenarios and treat warnings
+appropriately, as detailed in a proposal below. The key points can be summarized
+in two sentences:
+
+* **Preparing a package for uploading to Hackage (e.g. ``cabal sdist``) should
+  fail if any warnings are produced.**
```

Comment:
I disagree for two very practical reasons:

1. Cabal can produce sdists very quickly and doesn't require anything else from the environment. Checking for Haskell warnings adds a dependency on GHC and possibly system packages, and makes the process potentially very slow.
2. It's impossible to enforce from the Hackage side.

Reply:
Reasonable objections. But these are surmountable, if we want to make this change: `cabal` could retain some information saying whether the last build was warning-free (or even which warnings were produced) and then check the information during `sdist`. This would not induce a dependency on GHC. And similarly it could upload the evidence if we wanted a Hackage check (although I don't think the Hackage check is necessary).

I'm not arguing here that all of this is a good idea -- just that these implementation matters need not slow us down from the general idea.